### PR TITLE
FEAT-#6256: HDK: Add support for DataFrameGroupBy.head/tail()

### LINKED ISSUE #6256: HDK: Add support for DataFrameGroupBy.head()
Currently DataFrameGroupBy.head() is not supported by the HDK backend and defaults to pandas.

### PR DESCRIPTION
**Pull request by AndreyPavlenko**
## What do these changes do?

- [x] first commit message and PR title follow format outlined [here](https://modin.readthedocs.io/en/latest/development/contributing.html#commit-message-formatting)
  > **_NOTE:_**  If you edit the PR title to match this format, you need to add another commit (even if it's empty) or amend your last commit for the CI job that checks the PR title to pick up the new PR title.
- [x] passes `flake8 modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] passes `black --check modin/ asv_bench/benchmarks scripts/doc_checker.py`
- [x] signed commit with `git commit -s` 
- [x] Resolves #6256 
- [x] tests added and passing
- [ ] module layout described at `docs/development/architecture.rst` is up-to-date

### CODE REVIEW

**Review by ienkovich — DISMISSED** [review]

File: modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py [hunk 1]
```diff
@@ -1036,7 +1038,7 @@ class SortNode(DFAlgNode):
         Holds a single sorted frame.
     columns : list of str
         A list of key columns for a sort.
-    ascending : bool
+    ascending : list of bool
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py:1041]: list with one element?

Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py:1041]: In my understanding, it means a list containing elements of type bool. Should it be `list of bools`?

Comment (anmyachev) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py:1041]: no, I just missed that this parameter can be a list. Do you have a test for this case?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/calcite_serializer.py [hunk 2]
```diff
@@ -133,8 +133,10 @@ def serialize_item(self, item):
             return self.serialize_obj(item)
         elif isinstance(item, list):
             return [self.serialize_item(v) for v in item]
+        elif isinstance(item, dict):
+            return {k: self.serialize_item(v) for k, v in item.items()}
 
-        self.expect_one_of(item, str, int)
+        self.expect_one_of(item, str, int, type(None))
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/calcite_serializer.py:139]: I suppose the function can return None now? In this case, need to update the docstring.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 3]
```diff
@@ -550,6 +550,10 @@ def groupby_agg(self, by, axis, agg, groupby_args, **kwargs):
                 + f"met '{type(by_frame._op).__name__}'."
             )
 
+        if agg in ("head", "tail"):
+            n = kwargs["agg_kwargs"]["n"]
+            return self._groupby_head_tail(agg, n, groupby_cols.tolist())
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:555]: Why do this conversion?

```suggestion
            return self._groupby_head_tail(agg, n, groupby_cols)
```

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 4]
```diff
@@ -564,6 +564,30 @@ def copy(self):
...
+        if operands := getattr(self, "operands", None):
+            expr = self
+            for i, op in enumerate(operands):
+                new_op = yield op
+                if new_op is not None:
+                    if new_op is not op:
+                        if expr is self:
+                            expr = self.copy()
+                        expr.operands[i] = new_op
+                    yield expr
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:589]: Codecov writes that these lines are not covered by testing. This is true?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py [hunk 5]
```diff
@@ -636,6 +640,86 @@ def generate_by_name(by):
...
+            "order_key": 0,
+        }
+        row_num_op.upper_bound = {
+            "unbounded": False,
+            "preceding": False,
+            "following": False,
+            "is_current_row": True,
+            "offset": None,
+            "order_key": 1,
+        }
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py:703]: Can you move the initialization of this operation into a separate function?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 6]
```diff
@@ -1042,8 +1105,8 @@ def __repr__(self):
         -------
         str
         """
-        if len(self.operands) == 1:
-            return f"({self.op} {self.operands[0]} [{self._dtype}])"
+        if pk := getattr(self, "partition_keys", None):
+            return f"({self.op} {self.operands} {pk} {self.order_keys} [{self._dtype}])"
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:1109]: Is it not covered by tests?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 7]
```diff
@@ -944,7 +968,46 @@ def copy(self):
         -------
         OpExpr
         """
-        return OpExpr(self.op, self.operands.copy(), self._dtype)
+        op = OpExpr(self.op, self.operands.copy(), self._dtype)
+        if pk := getattr(self, "partition_keys", None):
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:972]: Do you have a description of what "partition_keys" field means somewhere?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 8]
```diff
@@ -944,7 +968,46 @@ def copy(self):
...
+    ) -> Generator[Type["BaseExpr"], Type["BaseExpr"], Type["OpExpr"]]:
+        expr = self
+        for i, op in enumerate(self.operands):
+            new_op = yield op
+            if new_op is not None:
+                if new_op is not op:
+                    if expr is self:
+                        expr = self.copy()
+                    expr.operands[i] = new_op
+                yield expr
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:992]: At least in two places there is a generator by `operands`, can it be taken out into a separate function?

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 9]
```diff
@@ -564,6 +564,30 @@ def copy(self):
         """
         pass
 
+    def nested_expressions(
+        self,
+    ) -> Generator[Type["BaseExpr"], Type["BaseExpr"], Type["BaseExpr"]]:
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:569]: "If your generator will only yield values, set the SendType and **ReturnType** to `None`:" https://docs.python.org/3/library/typing.html#typing.Generator

```suggestion
    ) -> Generator[Type["BaseExpr"], Type["BaseExpr"], None]:
```

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 10]
```diff
@@ -944,7 +968,46 @@ def copy(self):
...
+            op.is_rows = self.is_rows
+            op.order_keys = self.order_keys
+            op.lower_bound = self.lower_bound
+            op.upper_bound = self.upper_bound
+        return op
+
+    @_inherit_docstrings(BaseExpr.nested_expressions)
+    def nested_expressions(
+        self,
+    ) -> Generator[Type["BaseExpr"], Type["BaseExpr"], Type["OpExpr"]]:
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:983]:
```suggestion
    ) -> Generator[Type["BaseExpr"], Type["BaseExpr"], None]:
```

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py [hunk 11]
```diff
@@ -944,7 +968,46 @@ def copy(self):
...
+        if pk := getattr(self, "partition_keys", None):
+            op.partition_keys = pk
+            op.is_rows = self.is_rows
+            op.order_keys = self.order_keys
+            op.lower_bound = self.lower_bound
+            op.upper_bound = self.upper_bound
+        return op
+
+    @_inherit_docstrings(BaseExpr.nested_expressions)
+    def nested_expressions(
```
Comment (anmyachev) [inline modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:981]: In this function, there are 3 generators for three different entities at once. Please explain how it works and why it is done this way.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py (hunk 1 above)
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/df_algebra.py:1041]: Ah, sorry, didn't get the question. It should always be a list. No, there is no a special test case.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py (hunk 11 above)
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:981]: The generator is used not only for iterating over the nested expressions but also it allows to replace some expressions. The expressions are replaced not in the current object, but a copy is created. This functionality is used by `BaseExpr.translate_input()`, `BaseExpr.fold()` and `_maybe_copy_and_translate_expr()`.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py (hunk 7 above)
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:1023]: Added to the OpExpr attributes doc.

File: modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py (hunk 6 above)
Comment (AndreyPavlenko) [reply modin/experimental/core/execution/native/implementations/hdk_on_native/expr.py:1151]: No, this is a __repr__ method of an internal class, mainly used for debugging.

**Review by anmyachev — APPROVED** [review]

**Pull request merged**